File: python/src/agentlib/skills/tool.py
"""SkillTool — Tool Protocol implementation for loading skills."""

from __future__ import annotations

from ..tool import Tool, ToolContext, ToolResult
from ..types import ContentBlock, Message, Role, TextBlock
from .loader import SkillLoader
from .substitution import substitute_variables
from .types import SkillInfo


class SkillTool(Tool):
    """Tool that loads a skill and returns its processed content.

    Registered automatically when agent config has ``enable_skills=True``.
    """

    name = "skill"
    is_read_only = True

    input_schema = {
        "type": "object",
        "properties": {
            "skill": {
                "type": "string",
                "description": "The name of the skill to load",
            },
            "args": {
                "type": "string",
                "description": (
                    "Optional arguments passed to the skill via "
                    "$ARGUMENTS variable substitution"
                ),
            },
        },
        "required": ["skill"],
    }

    def __init__(self, loader: SkillLoader) -> None:
        self._loader = loader

    @property
    def description(self) -> str:
        desc = (
            "Load a skill and get its instructions. "
            "Skills provide specialized capabilities for specific tasks."
        )
        skills = self._loader.discover_all()
        invocable = [s for s in skills if s.metadata.user_invocable]
        if invocable:
            desc += "\n\nAvailable skills:\n"
            for skill in invocable:
                if skill.metadata.description:
                    desc += f"- {skill.metadata.name}: {skill.metadata.description}\n"
                else:
                    desc += f"- {skill.metadata.name}\n"
        return desc

    async def call(
        self, input: dict, context: ToolContext
    ) -> ToolResult:
        skill_name = input.get("skill", "")
        args = input.get("args")

        if not skill_name:
            return ToolResult.error(
                '"skill" is required'
            )

        skill = self._loader.find_by_name(skill_name)
        if not skill:
            available = self._loader.discover_all()
            names = [s.metadata.name for s in available]
            names_str = ", ".join(names) if names else "(none)"
            return ToolResult.error(
                f'Skill not found: "{skill_name}". '
                f"Available skills: {names_str}"
            )

        processed = substitute_variables(
            skill.content,
            args=args,
            skill_dir=skill.dir_path,
        )

        full_content = _build_skill_injection_content(skill, processed)
        brief = f"Skill loaded: {skill.metadata.name}"
        injection_msg = Message(
            role=Role.USER,
            content=[TextBlock(text=full_content)],
        )
        return ToolResult.success_with_messages(brief, [injection_msg])


def _build_skill_injection_content(skill: SkillInfo, content: str) -> str:
    """Format a skill's metadata and content into an injection string."""
    parts: list[str] = []
    parts.append(f"## Skill: {skill.metadata.name}")

    if skill.metadata.description:
        parts.append(f"Description: {skill.metadata.description}")
    if skill.metadata.when_to_use:
        parts.append(f"When to use: {skill.metadata.when_to_use}")

    parts.append("")
    parts.append(content)

    if skill.metadata.allowed_tools:
        parts.append("")
        parts.append(
            f"Note: When following this skill's instructions, "
            f"only use these tools: {', '.join(skill.metadata.allowed_tools)}"
        )

    if skill.metadata.context == "fork":
        parts.append("")
        parts.append("This skill should be executed in a fork context.")

    return "\n".join(parts)

File: go/skills/tool.go
package skills

import (
	"fmt"
	"strings"
)

// SkillTool provides the "skill" tool logic. It is not itself an agentlib.Tool
// (to avoid circular imports). The agent wraps it into the proper interface.
type SkillTool struct {
	loader *Loader
}

// NewSkillTool creates a new SkillTool backed by the given loader.
func NewSkillTool(loader *Loader) *SkillTool {
	return &SkillTool{loader: loader}
}

// ToolName returns the name of the tool.
func (t *SkillTool) ToolName() string {
	return "skill"
}

// ToolDescription returns the description of the tool, dynamically listing
// user-invocable skills.
func (t *SkillTool) ToolDescription() string {
	desc := "Load a skill and get its instructions. Skills provide specialized capabilities for specific tasks."
	skills := t.loader.DiscoverAll()
	var invocable []SkillInfo
	for _, s := range skills {
		if s.Metadata.UserInvocable {
			invocable = append(invocable, s)
		}
	}
	if len(invocable) > 0 {
		desc += "\n\nAvailable skills:\n"
		for _, skill := range invocable {
			if skill.Metadata.Description != "" {
				desc += fmt.Sprintf("- %s: %s\n", skill.Metadata.Name, skill.Metadata.Description)
			} else {
				desc += fmt.Sprintf("- %s\n", skill.Metadata.Name)
			}
		}
	}
	return desc
}

// ToolInputSchema returns the JSON schema for the tool input.
func (t *SkillTool) ToolInputSchema() map[string]any {
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"skill": map[string]any{
				"type":        "string",
				"description": "The name of the skill to load",
			},
			"args": map[string]any{
				"type":        "string",
				"description": "Optional arguments passed to the skill via $ARGUMENTS variable substitution",
			},
		},
		"required": []any{"skill"},
	}
}

// ToolIsReadOnly returns true since loading a skill is a read-only operation.
func (t *SkillTool) ToolIsReadOnly() bool {
	return true
}

// Execute loads a skill by name, substitutes variables, and returns the formatted result.
//
// Deprecated: Use ExecuteWithInjection instead.
func (t *SkillTool) Execute(name, args string) (string, error) {
	return t.ExecuteWithArgs(name, args)
}

// ExecuteWithArgs loads a skill by name, substitutes variables, and returns the formatted result.
func (t *SkillTool) ExecuteWithArgs(name, args string) (string, error) {
	if name == "" {
		return "", fmt.Errorf("\"skill\" is required")
	}

	skill := t.loader.FindByName(name)
	if skill == nil {
		all := t.loader.DiscoverAll()
		var names []string
		for _, s := range all {
			names = append(names, s.Metadata.Name)
		}
		available := "(none)"
		if len(names) > 0 {
			available = strings.Join(names, ", ")
		}
		return "", fmt.Errorf("Skill not found: %q. Available skills: %s", name, available)
	}

	processedContent := SubstituteVariables(skill.Content, args, skill.DirPath)
	return BuildSkillInjectionContent(skill, processedContent), nil
}

// ExecuteWithInjection loads a skill by name and returns a brief confirmation along
// with injection messages containing the full skill content. The injection messages
// will be merged into the conversation history so the model sees the full instructions
// as user context.
func (t *SkillTool) ExecuteWithInjection(name, args string) (string, []InjectionMessage, error) {
	if name == "" {
		return "", nil, fmt.Errorf("\"skill\" is required")
	}

	skill := t.loader.FindByName(name)
	if skill == nil {
		all := t.loader.DiscoverAll()
		var names []string
		for _, s := range all {
			names = append(names, s.Metadata.Name)
		}
		available := "(none)"
		if len(names) > 0 {
			available = strings.Join(names, ", ")
		}
		return "", nil, fmt.Errorf("Skill not found: %q. Available skills: %s", name, available)
	}

	processedContent := SubstituteVariables(skill.Content, args, skill.DirPath)
	fullContent := BuildSkillInjectionContent(skill, processedContent)
	brief := fmt.Sprintf("Skill loaded: %s", skill.Metadata.Name)
	injectionMsg := InjectionMessage{
		Role:    "user",
		Content: fullContent,
	}
	return brief, []InjectionMessage{injectionMsg}, nil
}

// InjectionMessage represents a message to be injected into the conversation
// history alongside tool results.
type InjectionMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// BuildSkillInjectionContent formats a skill's metadata and processed content
// into a full instruction string for injection into the conversation.
// It includes allowed_tools hints and fork mode markers.
func BuildSkillInjectionContent(skill *SkillInfo, content string) string {
	var parts []string
	parts = append(parts, fmt.Sprintf("## Skill: %s", skill.Metadata.Name))

	if skill.Metadata.Description != "" {
		parts = append(parts, fmt.Sprintf("Description: %s", skill.Metadata.Description))
	}
	if skill.Metadata.WhenToUse != "" {
		parts = append(parts, fmt.Sprintf("When to use: %s", skill.Metadata.WhenToUse))
	}

	parts = append(parts, "")
	parts = append(parts, content)

	if len(skill.Metadata.AllowedTools) > 0 {
		parts = append(parts, "")
		parts = append(parts, fmt.Sprintf("Note: When following this skill's instructions, only use these tools: %s", strings.Join(skill.Metadata.AllowedTools, ", ")))
	}

	if skill.Metadata.Context == "fork" {
		parts = append(parts, "")
		parts = append(parts, "This skill should be executed in a fork context.")
	}

	return strings.Join(parts, "\n")
}
